### ISSUE TITLE
v12 Login image not displaying

### ISSUE THREAD

**Issue by sempervictus** [nextcloud/nextcloud-snap#394]
After upgrading from 11->12, the login background vanished.
Uploaded again through settings, saved message appears, but the progress circle next to the "saved" green indicator continues to churn, and the login page retains the default NextCloud image.

**Comment by sempervictus** [comment]
Logs show repeating patterns of:
```

Error | lib | ResourceLocator  can not find a web root (root:  /snap/nextcloud/4371/htdocs/apps/files_texteditor, file:  index.php/css/files_texteditor/4c4f06b6d3abda5a726d934e5991d3d6-merged.css,  webRoot: , throw: true) | 2017-12-12T23:00:08-0500
-- | -- | -- | --
Error | lib | ResourceLocator  can not find a web root (root:  /snap/nextcloud/4371/htdocs/apps/files_sharing, file:  index.php/css/files_sharing/4c4f06b6d3abda5a726d934e5991d3d6-mergedAdditionalStyles.css,  webRoot: , throw: true) | 2017-12-12T23:00:08-0500
Error | lib | ResourceLocator  can not find a web root (root: /snap/nextcloud/4371/htdocs/apps/files,  file: index.php/css/files/4c4f06b6d3abda5a726d934e5991d3d6-merged.css,  webRoot: , throw: true) | 2017-12-12T23:00:08-0500

```
along with several instances of:
```

Error | no app in context | OCP\AppFramework\QueryException: Could not resolve OCA\Spreed\Settings\Admin! Class OCA\Spreed\Settings\Admin does not exist/var/lib/snapd/snap/nextcloud/4371/htdocs/lib/private/AppFramework/Utility/SimpleContainer.php - line 117:  OC\AppFramework\Utility\SimpleContainer->resolve('OCA\\Spreed\\Sett...')/var/lib/snapd/snap/nextcloud/4371/htdocs/lib/private/ServerContainer.php - line 132:  OC\AppFramework\Utility\SimpleContainer->query('OCA\\Spreed\\Sett...')/var/lib/snapd/snap/nextcloud/4371/htdocs/lib/private/Settings/Manager.php - line 261:  OC\ServerContainer->query('OCA\\Spreed\\Sett...')/var/lib/snapd/snap/nextcloud/4371/htdocs/lib/private/Settings/Manager.php - line 352:  OC\Settings\Manager->query('OCA\\Spreed\\Sett...')/var/lib/snapd/snap/nextcloud/4371/htdocs/settings/Controller/AdminSettingsController.php - line 138:  OC\Settings\Manager->getAdminSettings('additional')/var/lib/snapd/snap/nextcloud/4371/htdocs/settings/Controller/AdminSettingsController.php - line 72:  OC\Settings\Controller\AdminSettingsController->getNavigationParameters('theming')[internal function] OC\Settings\Controller\AdminSettingsController->index('theming')/var/lib/snapd/snap/nextcloud/4371/htdocs/lib/private/AppFramework/Http/Dispatcher.php - line 160:  call_user_func_array(Array, Array)/var/lib/snapd/snap/nextcloud/4371/htdocs/lib/private/AppFramework/Http/Dispatcher.php - line 90:  OC\AppFramework\Http\Dispatcher->executeController(Object(OC\Settings\Controller\AdminSettingsController), 'index')/var/lib/snapd/snap/nextcloud/4371/htdocs/lib/private/AppFramework/App.php - line 114:  OC\AppFramework\Http\Dispatcher->dispatch(Object(OC\Settings\Controller\AdminSettingsController), 'index')/var/lib/snapd/snap/nextcloud/4371/htdocs/lib/private/AppFramework/Routing/RouteActionHandler.php - line 47:  OC\AppFramework\App  main('OC\\Settings\\Con...', 'index', Object(OC\AppFramework\DependencyInjection\DIContainer), Array)[internal function] OC\AppFramework\Routing\RouteActionHandler->__invoke(Array)/var/lib/snapd/snap/nextcloud/4371/htdocs/lib/private/Route/Router.php - line 299:  call_user_func(Object(OC\AppFramework\Routing\RouteActionHandler), Array)/var/lib/snapd/snap/nextcloud/4371/htdocs/lib/base.php - line 1004:  OC\Route\Router->match('/settings/admin...')/var/lib/snapd/snap/nextcloud/4371/htdocs/index.php - line 48:  OC  handleRequest(){main} | 2017-12-12T22:55:49-0500
Error | no app in context | OCP\AppFramework\QueryException: Could not resolve OCA\OJSXC\Settings\Section! Class OCA\OJSXC\Settings\Section does not exist/var/lib/snapd/snap/nextcloud/4371/htdocs/lib/private/AppFramework/Utility/SimpleContainer.php - line 117:  OC\AppFramework\Utility\SimpleContainer->resolve('OCA\\OJSXC\\Setti...')/var/lib/snapd/snap/nextcloud/4371/htdocs/lib/private/ServerContainer.php - line 132:  OC\AppFramework\Utility\SimpleContainer->query('OCA\\OJSXC\\Setti...')/var/lib/snapd/snap/nextcloud/4371/htdocs/lib/private/Settings/Manager.php - line 261:  OC\ServerContainer->query('OCA\\OJSXC\\Setti...')/var/lib/snapd/snap/nextcloud/4371/htdocs/lib/private/Settings/Manager.php - line 289:  OC\Settings\Manager->query('OCA\\OJSXC\\Setti...')/var/lib/snapd/snap/nextcloud/4371/htdocs/settings/Controller/AdminSettingsController.php - line 133:  OC\Settings\Manager->getAdminSections()/var/lib/snapd/snap/nextcloud/4371/htdocs/settings/Controller/AdminSettingsController.php - line 72:  OC\Settings\Controller\AdminSettingsController->getNavigationParameters('theming')[internal function] OC\Settings\Controller\AdminSettingsController->index('theming')/var/lib/snapd/snap/nextcloud/4371/htdocs/lib/private/AppFramework/Http/Dispatcher.php - line 160:  call_user_func_array(Array, Array)/var/lib/snapd/snap/nextcloud/4371/htdocs/lib/private/AppFramework/Http/Dispatcher.php - line 90:  OC\AppFramework\Http\Dispatcher->executeController(Object(OC\Settings\Controller\AdminSettingsController), 'index')/var/lib/snapd/snap/nextcloud/4371/htdocs/lib/private/AppFramework/App.php - line 114:  OC\AppFramework\Http\Dispatcher->dispatch(Object(OC\Settings\Controller\AdminSettingsController), 'index')/var/lib/snapd/snap/nextcloud/4371/htdocs/lib/private/AppFramework/Routing/RouteActionHandler.php - line 47:  OC\AppFramework\App  main('OC\\Settings\\Con...', 'index', Object(OC\AppFramework\DependencyInjection\DIContainer), Array)[internal function] OC\AppFramework\Routing\RouteActionHandler->__invoke(Array)/var/lib/snapd/snap/nextcloud/4371/htdocs/lib/private/Route/Router.php - line 299:  call_user_func(Object(OC\AppFramework\Routing\RouteActionHandler), Array)/var/lib/snapd/snap/nextcloud/4371/htdocs/lib/base.php - line 1004:  OC\Route\Router->match('/settings/admin...')/var/lib/snapd/snap/nextcloud/4371/htdocs/index.php - line 48:  OC  handleRequest(){main} | 2017-12-12T22:55:49-0500
Error | no app in context | OCP\AppFramework\QueryException: Could not resolve OCA\IssueTemplate\Settings\Section! Class OCA\IssueTemplate\Settings\Section does not exist/var/lib/snapd/snap/nextcloud/4371/htdocs/lib/private/AppFramework/Utility/SimpleContainer.php - line 117:  OC\AppFramework\Utility\SimpleContainer->resolve('OCA\\IssueTempla...')/var/lib/snapd/snap/nextcloud/4371/htdocs/lib/private/ServerContainer.php - line 132:  OC\AppFramework\Utility\SimpleContainer->query('OCA\\IssueTempla...')/var/lib/snapd/snap/nextcloud/4371/htdocs/lib/private/Settings/Manager.php - line 261:  OC\ServerContainer->query('OCA\\IssueTempla...')/var/lib/snapd/snap/nextcloud/4371/htdocs/lib/private/Settings/Manager.php - line 289:  OC\Settings\Manager->query('OCA\\IssueTempla...')/var/lib/snapd/snap/nextcloud/4371/htdocs/settings/Controller/AdminSettingsController.php - line 133:  OC\Settings\Manager->getAdminSections()/var/lib/snapd/snap/nextcloud/4371/htdocs/settings/Controller/AdminSettingsController.php - line 72:  OC\Settings\Controller\AdminSettingsController->getNavigationParameters('theming')[internal function] OC\Settings\Controller\AdminSettingsController->index('theming')/var/lib/snapd/snap/nextcloud/4371/htdocs/lib/private/AppFramework/Http/Dispatcher.php - line 160:  call_user_func_array(Array, Array)/var/lib/snapd/snap/nextcloud/4371/htdocs/lib/private/AppFramework/Http/Dispatcher.php - line 90:  OC\AppFramework\Http\Dispatcher->executeController(Object(OC\Settings\Controller\AdminSettingsController), 'index')/var/lib/snapd/snap/nextcloud/4371/htdocs/lib/private/AppFramework/App.php - line 114:  OC\AppFramework\Http\Dispatcher->dispatch(Object(OC\Settings\Controller\AdminSettingsController), 'index')/var/lib/snapd/snap/nextcloud/4371/htdocs/lib/private/AppFramework/Routing/RouteActionHandler.php - line 47:  OC\AppFramework\App  main('OC\\Settings\\Con...', 'index', Object(OC\AppFramework\DependencyInjection\DIContainer), Array)[internal function] OC\AppFramework\Routing\RouteActionHandler->__invoke(Array)/var/lib/snapd/snap/nextcloud/4371/htdocs/lib/private/Route/Router.php - line 299:  call_user_func(Object(OC\AppFramework\Routing\RouteActionHandler), Array)/var/lib/snapd/snap/nextcloud/4371/htdocs/lib/base.php - line 1004:  OC\Route\Router->match('/settings/admin...')/var/lib/snapd/snap/nextcloud/4371/htdocs/index.php - line 48:  OC  handleRequest(){main} | 2017-12-12T22:55:49-0500

```

**Comment by kyrofa** [comment]
So this is a custom background image? Are you using the theming app? Those errors are very much reminiscent of the myriad of CSS regressions introduced in v12 that caused us to hold back the update for so long. Assuming this is the theming app, looks like it (or a code path used by it) also has that issue. If you're able to give steps to reproduce from a clean install, that would be super helpful.

**Comment by sempervictus** [comment]
This was done by installing the snap at v11 and enabling the theming app, uploading an image and running that way until the v12 rollover. At that point, the app still indicates that an image was uploaded, but the default one is presented. Additional attempts to re-upload the image to replace the default seem to succeed, but the preview generation never completes and the actual image is never updated.
Im concerned, given the use of the theming app (and not a core function) that its not so much a clean install issue since the version rollover carries existing users and data (along with any customizations made via the functionality of the original snap).

**Comment by kyrofa** [comment]
Oh I'm not concerned that it's an issue with a clean install either, I just need a way to reproduce :) .

**Comment by pherzing** [comment]
I can corroborate sempervictus's notes: 11.x snap nextcloud, upgraded to 12.0.4 this morning, same behavior re theming. For what it's worth, Safari 11 is saying:

> Did not parse stylesheet at 'https://nextcloud.companyname.com/index.php/apps/theming/theming.scss' because non CSS MIME types are not allowed in strict mode.

Firefox 57 makes a similar comment:

> Style sheet could not be loaded https://nextcloud.companyname.com/index.php/apps/theming/theming.scss

**Comment by ti-guru** [comment]
I can confirm the issue as my instance was also upgraded from 11 to 12. Seems like the issue affects choice of color, logo and background picture. Also, the preview thumbnail of the logo has moved and the login thumbnail does not generate.
![theming_app](https://user-images.githubusercontent.com/31796171/33951997-ccf6269e-e030-11e7-9406-bcd679a811f5.png)
![login](https://user-images.githubusercontent.com/31796171/33952005-d18e3e80-e030-11e7-92fb-712579a915ee.png)

**Comment by kyrofa** [comment]
I think I've got enough info, I'll play with it when I'm able. However, I'm pretty certain this is an upstream Nextcloud issue, so I doubt we can fix it here. If you need theming, it sounds like you may have to revert and stay on v11 until v12 fixes this regression. To do that, you first need to revert back to v11:

    $ sudo snap revert nextcloud
    nextcloud reverted to 11.0.5snap2

Then refresh to the `11/stable` channel to make sure you don't get v12 updates:

    kyrofa@Pandora:~$ sudo snap refresh nextcloud --channel=11/stable
    nextcloud (11/stable) 11.0.5snap2 from 'nextcloud' refreshed
    kyrofa@Pandora:~$ snap info nextcloud
    <snip>
    tracking:                   11/stable
    <snip>

**Comment by pherzing** [comment]
I'm too poorly informed to understand the subtleties of the snap architecture, so I offer this only in case it's helpful for scoping the problem:

I operate a separate nextCloud 12.0.4 server that was built on a full stack (instead of the snap deployment) only a few days ago, because it needed to integrate with an Active Directory setup. Theming there is working as expected.

Again, I'm not disputing that it's a general nextCloud issue. The two servers are _primarily_ different in that one is snap and one is not, but many settings differ between them, and unlike my snap install, the full-stack system was not an upgrade from 11; it started its life at 12.0.4.

Again, only if it's helpful, but both are running on Ubuntu 16.0.4, on Digital Ocean droplets.

**Comment by kyrofa** [comment]
@pherzing right, I'm not trying to say that theming is generally broken in upstream Nextcloud, I'm trying to say that v12 changed the way assets in apps worked from v11. In the snap (which is read-only) we store apps installed by the user outside of the webroot, which is supported by Nextcloud, and worked great in v11, but with their v12 refactor broke in numerous places. We fixed all the ones we knew about, but it looks like there's another one in the theming app somewhere.

**Comment by pherzing** [comment]
Thanks for the response. It helps to understand the why, even if I can't fix it myself.

**Comment by WereCatf** [comment]
I keep seeing "ResourceLocator can not find a web root (root: /var/snap/nextcloud/4371/nextcloud/extra-apps/contacts, file: index.php/css/contacts/d7a9968d2ff19408599e6832f8b4b3d1-style.css, webRoot: , throw: true)" in my log-files whenever I open the contacts-tab in Nextcloud, and I do not know if it's related to this, but it seems to me like it just might be. I have never used theming with Nextcloud, though, or even tried it, so I have no idea why this happens. Doesn't seem to be causing me any issues, but I am just reporting this just in case it's related to this issue here and if my report might be of any help.

**Comment by SkyWheel** [comment]
Can confirm the same issue after upgrading from 11.x.x-stable to 12/edge long time ago (June 2017???).

**Comment by kyrofa** [comment]
@WereCatf that sounds like a different issue that I personally fixed. Does the contacts app not work properly? I'm using it here and it's working fine, no log errors. You might want to log a different issue.

**Comment by tobia** [comment]
I directly installed 12.0.4snap1 4371 on a fresh server, enabled https with `nextcloud.enable-https lets-encrypt` and went straight to Admin / Theming to change logo and background.

The logo appears correctly in the upper left corner when logged in, but the login page shows the default logo and background image. Also Loading preview... keeps spinning forever.

**Comment by tobia** [comment]
By the way, I just started using snaps. Is there a way to do a manual override? Say that I (root) wanted to personalize a file inside the snap package, what options do I have?

**Comment by sempervictus** [comment]
@tobia: I think you're talking about rolling your own snaps for that since by definition they are immutable once built.

**Comment by kyrofa** [comment]
@sempervictus is right, if you want to change, say, a .php file in `/snap/nextcloud/current/htdocs/`, you do indeed need to roll your own snap.

**Comment by dialed** [comment]
Hello. I am new to nextcloud and snaps but installed via digitalocean and really like the functionality. Is changing the nextcloud logo at the login screen a feature that is coming to the snap or must one install manually to get that functionality? Thank you in advance!

**Comment by kyrofa** [comment]
Hey @dialed! Indeed, the fact that it doesn't work is just a bug, we'll track it down.

**Comment by kyrofa** [comment]
As mentioned above, you can install v11 for now if you require this functionality from the 11/stable channel.

**Comment by dialed** [comment]
Thanks @kyrofa sorry about the confusion, total noob here!

**Comment by kyrofa** [comment]
Would you guys mind taking version 12.0.5 for a spin to see if anything has been fixed, there? Just install it on a test machine or something and see if it works the way you expect:

    $ sudo snap install nextcloud --channel=stable/pr-432

**Comment by tobia** [comment]
I just did this:

    # snap refresh nextcloud --channel=stable/pr-432
    nextcloud (stable/pr-432) 12.0.4snap2 from 'nextcloud' refreshed

And `/settings/admin` says 12.0.5

But the issue is still there. Loading a logo or a background image in `/settings/admin/theming` does not give an endless spinner as before, but neither the background nor the logo are shown on the login page.

I did empty my cache and reload a few times. The body of the login page seems to have a `background-image: url(../img/background.jpg?v=1)` coming from `guest.css?v=416fcc3f9f4e5df41ca45d3600d75c61-29` that still points to the default image, when opened in a new tab.

**Comment by QuaySo** [comment]
I tried this on a fresh server and had the exact same results as @tobia mentioned. Strange enough the uploaded background image is not visible in the webclient, but in the Android app it is!

**Comment by tobia** [comment]
@QuaySo this proves that the images are saved correctly, so there must be a (second?) bug specific to the web UI, that has to do with the immutability of the snap package.

**Comment by kyrofa** [comment]
> there must be a (second?) bug specific to the web UI, that has to do with the immutability of the snap package.

More specifically a bug that has something to do with storing these assets out of the webroot.

**Comment by kyrofa** [comment]
Thank you both @tobia and @QuaySo for testing it out, that's frustrating.

**Comment by QuaySo** [comment]
I did some extra testing and noticed that uploading a new background image made a change in this folder: 

/var/snap/nextcloud/common/nextcloud/data/appdata_ocnvdjc7dnm0/theming/images

The filename is just "background", no extension. It's not a 1:1 copy of the image that I uploaded, but I guess Nextcloud is processing it before saving the file. I checked the file and it displays the correct image when I added a png extension.

@kyrofa: Is this info helpful? I could do some more testing if needed...

**Comment by kyrofa** [comment]
Wait, to clarify: if you name it `background.png` instead of `background` it works as expected?

**Comment by QuaySo** [comment]
Sorry for being not clear. I added the extension to be able to see a preview and confirm the file is really the image that I uploaded. Just that! 
It didn't solve the issue.

**Comment by metheis** [comment]
Sorry, I may be late to this thread, but I did a fresh ubuntu-server 16.04.3 installation a couple of days ago, and I installed Nexcloud 13 (beta at the time). The logo and background image is still definitely a problem. When I upload a logo, you can see it (very small) in the top left corner when logged in. However, the logo nor the background image appear in the login page. 

Something interesting to note, is that in the login page, if you try and drag the logo (as if you were dragging an image off google), I can see the translucent version of the uploaded logo. The next cloud logo is the actual logo on the page, but it's interesting the (proper) uploaded one appears when trying to drag the image.

**Comment by newmediologo** [comment]
I've a similar issue.. Changed the logo, on the login page I always see the default logo but if I go on the logo and click on "Open image in another tab" appear my logo.. It's weird..

**Comment by kyrofa** [comment]
Thank you everyone for commenting on this issue. It's nice to see how many people utilize this feature! This bug is next on my list.

**Comment by sempervictus** [comment]
This issue is 2 months old, indicating that the snaps being shipped are not considered a first order member of the ecosystem. The issues posted here by all these users are not being addressed by core developers. 
Are snaps supported by nextcloud or is this issues section just for us to vent? What happens if we start posting security issues here? Zero days will just hang out in issues? We put some users on snaps for the supposed convenience and the fact that it shouldn't require support... Now everyone looks like an ass and I'm having to shift people to docker with all its idiocy because snaps perpetually don't work as their docker or manual install counterparts. Snaps should make this end user facing tech accessible to them, meaning "have no more bugs than any other release format" because end users will rip their hair out not understanding what an RO mount is or any of that.
How is next cloud allocating resources to problems in this issue tracker?

**Comment by tobia** [comment]
@sempervictus Woah, chill out there buddy. I don't remember this being a commercial support ticket system. If you have a support contract with somebody, you should complain to the people you paid money to. Otherwise you're just looking out of place here.

**Comment by sempervictus** [comment]
We don't do this sort of thing for commercial clients, too many security concerns around maintenance of a system which users might accidentally use to store phi or something else governed by regulatory mechanisms... That said, for humans wanting to not give their data to google or msft it fits the bill. Problem is that its marketed as a turnkey solution for that user base, and guess whom they harass when it doesn't work? No support contract, just time lost, because I'm not going to ignore non billable people out of basic decency... Think of setting this up for grandma and grandpa. Update happens, stuff looks different, they freak out...
Next cloud was supposed to not be repeating the confusing mess of own cloud, so my frustration mainly stems from the fact that its versioned as a v12 product without fully validating its packaging mechanisms as all working the same

**Comment by tobia** [comment]
@sempervictus Ok, I get it. In fact, snaps are good as an immutable application container, and I agree, an improvement over Docker, but they do require thorough testing and QC, because developers obviously aren't using them as their day-to-day platform.

But this is technically a minor issue, albeit an obscure one (because, as @metheis mentioned, there's some weird CSS shenanigans going on in the login page) and @kyrofa just said he's gonna look into it. Let's just hope he can get at the root of the issue.

I've actually seen people type their login credentials in the wrong, phishing website, before I could stop them, and not even understanding what they had done. So I understand the importance of having a unique looking login page.

**Comment by SkyWheel** [comment]
This is a really old issue. I have it since I've upgraded my Nextcloud snap to12/Edge channel last May.

**Comment by kyrofa** [comment]
> [...] versioned as a v12 product without fully validating its packaging mechanisms as all working the same

To the length of my knowledge, Nextcloud itself doesn't package anything. They simply make tarballs, and their community takes it from there. This is exactly that-- @pachulo and I (the main maintainers here) don't work for Nextcloud, we've just been a part of the community forever and we take care of the snap. We do this because we use it ourselves.

I share your frustration that things break between Nextcloud releases. This is why we delayed releasing v12 for so long: stuff broke between v11 and v12, and no one upstream seemed to care. I eventually burned an entire weekend delving into their code to figure out what broke, and made several pull requests to fix it myself. Unfortunately, I can't test every use-case, and this is an example of one that I personally missed. I'm sorry for that, I did my best.

I will point out, however, that v12 snaps were available far before they were actually released for everyone. You could have been testing this with us and helped us catch it ahead of time. A breakdown of the channels/releases available is [here](https://github.com/nextcloud/nextcloud-snap/wiki/Release-strategy). As we approached the stable release, I made [calls for testing](https://help.nextcloud.com/t/call-for-testing-nextcloud-snap-v12-0-4snap1/24524) in [multiple](https://forum.snapcraft.io/t/call-for-testing-nextcloud-snap-v12-0-4snap1/3102) [places](https://twitter.com/rainveil/status/939264992141963264).

I will not take responsibility for the fact that upstream Nextcloud has [roughly 50% test coverage](https://codecov.io/gh/nextcloud/server), which is how all these breakages occur in the first place. Every time I fix one of their problems, I try to enshrine the behavior as much as I can in tests. And you know what? In the last 6 months (according to that page), the coverage has gone _down_. As long as they continue to work with such terrible coverage, we will continue to see these types of regressions. Such is life.

In conclusion, I'm sorry this regression happened, and I'm sorry it has taken so long to get fixed, but it tends to take some convincing to get me to burn a weekend away from my wife and kids to fix a problem I didn't cause. Ranting about it doesn't help 😃 .

**Comment by sempervictus** [comment]
Ranting doesnt help, though it does expose the lil gem about this being an unsupported platform (at least by nextcloud staff).
Upstream testing does seem to have some holes in it, so maybe bringing the snap into the official fold would kill two birds - force testing improvement and packaging consistency. Otherwise theres no guarantee that they won't just break snaps permanently one day leaving everyone in the lurch who uses it.

**Comment by jasper321** [comment]
I have the same issue with the Snappy installation. If you install Nextcloud natively on your Ubuntu installation, it is working fine. Maybe it's a snappy problem?

**Comment by spot88** [comment]
Have this problem since the start, allways used snap installation. 
Seems to work fine with normal installation on Ubuntu 16.04.

I'm on version 13.0.0 now, and problem is still there. 
The logo seems to be there when using the mobile app (android). - Totally bugged size tought.

**Comment by kyrofa** [comment]
I've been investigating this. It definitely seems to have something to do with SCSS (just like the other bugs we fixed). I can trace it back to [this logic](https://github.com/nextcloud/server/blob/master/apps/theming/js/settings-admin.js#L53) which attempts to reload an SCSS stylesheet, where if I visit the URL generated [here](https://github.com/nextcloud/server/blob/master/apps/theming/js/settings-admin.js#L46) in normal Nextcloud it's compiled CSS, but if I visit it in the snap it's a 404. If I move the apps into another dir, in normaly Nextcloud it's uncompiled SCSS saying `@import 'theming.scss';`.

**Comment by sempervictus** [comment]
Is it feasible to force AOT compilation of all assets and "burn them into" the resulting squashfs image? If it has to generate runtime assets, do we need a snapcraft-specific patch for the paths where these go and are accessed from?

**Comment by kyrofa** [comment]
Nah, Nextcloud does it on the fly already; it writes them into the data dir. When you install a new app that uses SCSS, it'll compile those as well. For some reason it's just not compiling this one, not sure why yet.

**Comment by kyrofa** [comment]
Ah here, I think the problem is [here](https://github.com/nextcloud/server/blob/master/apps/theming/lib/Controller/ThemingController.php#L412), it's relying on the webroot.

**Comment by kyrofa** [comment]
Yep, that was it. Spinning the fix now, will have something for y'all to test soon.

**Comment by kyrofa** [comment]
Logged https://github.com/nextcloud/server/issues/8462 for your reference.

**Comment by kyrofa** [comment]
Alright, pull request is up: https://github.com/nextcloud/server/pull/8463.

**Comment by kyrofa** [comment]
I'm about to go to bed, but I wanted to get a snap up for folks to test before I did. Unfortunately the armhf and arm64 builds aren't done yet (I'll get them up in the morning), but amd64, i386, and ppc64 builds are up. Anyone bitten by this issue, please give the fix a test by using the `stable/394-theming-fix` channel:

    $ sudo snap install nextcloud --channel=stable/394-theming-fix

Or if you already have it installed:

    $ sudo snap refresh nextcloud --channel=stable/394-theming-fix

I'll comment again when the arm builds are up.

**Comment by QuaySo** [comment]
I tested it on a fresh install and both (logo and background) can be changed. Works as expected.
Nice fix!

**Comment by kyrofa** [comment]
Alright, the armhf and arm64 builds are up as well.

Thanks for testing, @QuaySo!

**Comment by tobia** [comment]
@kyrofa I can confirm that the issue is fixed also in the case of upgrade of an existing snap. Thank you!

**Comment by kyrofa** [comment]
Excellent, thanks @tobia!

**Comment by abarfety** [comment]
@kyrofa Issue is fixed, but it seems 394 broke Unsplash, app that shows a random photo on log in screen. With fix, no image appears, when reverted to 12.0.5snap1, Unsplash works again.

**Comment by kyrofa** [comment]
Wha...? Unsplash must somehow rely on Theming, then. This is definitely a fix for theming, though (the logic there was broken); sounds like Unsplash also needs a fix. I'll take a look when I'm able.

**Comment by abarfety** [comment]
To be more specific hoping to save you some time: Unsplash breaks after logo has been modified. I kinda found a weird fix: if you upload a background image then Unsplash works again, showing a random photo on log in screen (but not the image uploaded).

**Comment by kyrofa** [comment]
Oh, that sounds like a bug in Unsplash being exposed by theming actually working, now. Any chance you've tested this outside of the snap?

**Comment by abarfety** [comment]
Tested on v13.0.0, outside of the snap. Unsplash (with theming) worked as expected.

**Comment by innoliam** [comment]
`$ sudo snap refresh nextcloud --channel=stable/394-theming-fix`
Unfortunately, while the progress spinner now completes, I end up with a solid colour as BG instead of the PNG/JPG I uploaded (tried multiple). Is there a system dependency I'm missing (I manually pulled in php-imagemagick on Ubuntu 16.04)?

**Comment by kyrofa** [comment]
No, the snap includes all it's dependencies. Have you also set theme colors, or are you still using the default colors?

**Comment by kyrofa** [comment]
Note that this fix has landed in master, stable13, and stable12, and should be included in both 12.0.6 as well as 13.0.1. I'll close this issue once they're released and once we update.

**Comment by kyrofa** [comment]
You can see that this is fixed upstream by checking out the daily v12 snap for today (it'll give you a taste of 12.0.6):

    $ sudo snap install nextcloud --channel=12/edge

Or, if you already have it installed:

    $ sudo snap refresh nextcloud --channel=12/edge

**Comment by spot88** [comment]
Downgrading is not supported and is likely to cause unpredictable issues (from 13.0.0.14 to 12.0.5.3)

**Comment by kyrofa** [comment]
Yeah don't do that if you're on v13. Use `13/edge` instead to test the fix, there.

**Comment by lenusch** [comment]
Theming not working on Chrome with newest snap 13.0.1 rc1

**Comment by gcommit** [comment]
I have the same issue on 13.0.1 
As a workaround you can install the app "Unsplash" and deactivate it. For some reason this triggers a update of the login screen (do it that way, if you don't want to do changes on system itself)

**Comment by SkyWheel** [comment]
The trick with "Unsplash" didn't work for me :(

**Comment by gcommit** [comment]
Hey
You have to empty your browser cache as well. Or just use a private window. :)

**Comment by SkyWheel** [comment]
I was using private window. The Unsplash app worked as supposed, it was changing backgrounds. But it didn't make any difference for my custom theming, neither while the was turned on, nor while it disabled :(

**Comment by kyrofa** [comment]
Note that theming seems to be working as of 13.0.2snap1. I'm going to go ahead and close this issue, please let me know if there are still issues.
